Add consul module

acidmoose opened the pull request:
- Singleton pattern so that we can make coordination modules, like
a module that has dependencies to consul _and_ webconsole, and
can register the webconsole in consul.
- Includes re-registering if consul connection drops, to make sure
the service is registered.
- Includes deregistering on kill signals.

Commit:
Add consul module

Commit:
Add consul example

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,217 @@
+    public ConsulModule() {}
+
+    // private constructor. use getInstance().
+    private ConsulModule(final String host) {
```
If the consul agent object is injected here it gets very easy to write tests for this. So I would strongly recommend doing that.

acidmoose replied:
Sounds good.

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    /**
+     * Used by SecondBase.
```
Can you explain this more? Currently I'm not sure I get it. Also, this class is not a singleton or am I mistaken? One can call this constructor multiple times.

Also fetch the java book for the ultimate, Joshua Bloch approved, singleton pattern (backed by an enum).

acidmoose replied:
The class should be a singleton. This is the constructor used by secondbase core when doing the serviceloader things. I'm not sure how to get that working yet. I'll look into it again.

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,217 @@
+     * Deregister a list of services. To be used for deregistering consul services when
+     * shutting down.
+     */
+    private static class DeregisterThread extends Thread {
```
Why is this whole class embedded in this class while the four lines of config are in a separate one?

acidmoose replied:
It's just functionality of the class, really. And Flags are shared between modules. But I can move it out to.

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,217 @@
+     * shutting down.
+     */
+    private static class DeregisterThread extends Thread {
+        private Map<String, AgentClient> services = new HashMap<>();
```
This surely can be final, no?

acidmoose replied:
yes indeed

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,217 @@
+        private Map<String, AgentClient> services = new HashMap<>();
+
+        /**
+         * Record a serviceid to be deregistered at shutdown
```
Use either serviceId or "service id?"

acidmoose replied:
done

neumayer reviewed (commented):
Don't the oracle docs on loadable modules recommend to use singletons for all modules? Should we do that for all modules?

acidmoose commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,217 @@
+    public ConsulModule() {}
+
+    // private constructor. use getInstance().
+    private ConsulModule(final String host) {
```
I think this singleton thing is a very nice pattern to use. It makes it very easy to use for developers that wants to do other things through the modules, like registering their own service. ConsulModule.getInstance().registerService(...).

Yes, I think we should do this for all our modules.

acidmoose commented:
I think this singleton thing is a very nice pattern to use. It makes it very easy to use for developers that wants to do other things through the modules, like registering their own service. ConsulModule.getInstance().registerService(...).

Yes, I think we should do this for all our modules.

Commit:
Deregister thread new class. Description to constructor. Implement missing method.

neumayer commented:
Can you rebase this now and then we see if everything still makes sense?

Commit:
Add consul integration test

Commit:
Rebase and add full example

Commit:
Don't use singleton pattern in example.

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    /**
+     * Instatiate a ConsulModule.
```
Missing n.

acidmoose replied:
done

neumayer commented on example/src/main/java/org/secondbase/example/main/HelloAll.java:
```diff
@@ -0,0 +1,49 @@
+    public static void main(final String[] args) throws SecondBaseException, IOException {
+        final Counter mycounter = Counter.build("mycounter", "a counter").register();
+
+        final Widget prometheusWidget = new PrometheusWebConsole();
```
Use logging too?

        final SecondBaseModule jsonLoggerModule = new JsonLoggerModule();
        final Widget prometheusWidget = new PrometheusWebConsole();
        final Widget[] widgets = {prometheusWidget};
        final HttpWebConsole webconsole = new HttpWebConsole(widgets);
        final ConsulModule consul = new ConsulModule(ConsulModule.createLocalhostConsulClient());
        final SecondBaseModule[] modules = {jsonLoggerModule, webconsole, consul};

Use the interface type for the modules?

acidmoose replied:
done

acidmoose replied:
actually, I can't use the interface if I need to use the methods in the class. For instance, registering a service in consul.

neumayer commented on pom.xml:
```diff
@@ -21,6 +21,7 @@
         <module>webconsole</module>
         <module>jsonlogging</module>
         <module>prometheus-webconsole</module>
+        <module>consul</module>
```
This module transitively depends on guava 19.0. The json logging module already depends on guava 22.0. The orbitz consul thing does not work with guava 22.0. The solution for now should be to downgrade the guava version we use, put it in the parent pom and thus only manage the version once (possibly as a managed dependency), and make an upstream pull request for orbitz to work with guava 22.0.

neumayer replied:
This comment should be in the first pom file on the top.

acidmoose replied:
Added explicit dependency in consul and logging pom, version managed by the root pom.

neumayer replied:
Nice.

neumayer commented on example/src/main/java/org/secondbase/example/main/HelloConsul.java:
```diff
@@ -0,0 +1,32 @@
+        final String[] tags = {"tagone", "tagtwo"};
+
+        // Set consul endpoint.
+        ConsulModuleConfiguration.host = "localhost:8500";
```
Should we keep the examples so they all parse parameters, i.e. suply this as --consul-host in the args variable?
Now they're different, some use the args, some use the static fields directly.

acidmoose replied:
Yes, good point.

acidmoose replied:
I'm doing that in HelloAll now, for all flags. I'll keep this here as "optional use case".

neumayer commented on example/src/main/java/org/secondbase/example/main/HelloConsul.java:
```diff
@@ -0,0 +1,32 @@
+        // Set consul endpoint.
+        ConsulModuleConfiguration.host = "localhost:8500";
+
+        // Manually register a service in consul.
```
Remove the dot?

acidmoose replied:
done

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModuleConfiguration.java:
```diff
@@ -0,0 +1,16 @@
+     * Location of Consul server.
+     */
+    @Flag (
+            name="consul",
```
Call consul-host or consul-server?

acidmoose replied:
done

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModuleConfiguration.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Configuration parameters for the WebConsole.
+ */
+public final class ConsulModuleConfiguration {
```
Maybe we can come up with a rule we only have this separate class if there's more than five parameters? We could save quite some space by reintegrating this in the module classes everywhere. But we can do this later.

acidmoose replied:
agreed

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,190 @@
+     * @param healthCheckPath The api path which consul can use for health checks.
+     * @param healthCheckIntervalSec The interval between consul health checks.
+     * @param consulCustomTags Optionally register more consul tags with the service.
+     * @throws SecondBaseException
```
Add when it is thrown.

acidmoose replied:
done

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,190 @@
+     * @param healthCheckPath        The api path which consul can use for health checks.
+     * @param healthCheckIntervalSec The interval between consul health checks.
+     * @param consulCustomTags       Optionally register more consul tags with the service.
+     * @throws SecondBaseException
```
Add a word on in what cases it is thrown.

acidmoose replied:
done

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    /**
+     * Create a Consul connection to a localhost agent on port 8500.
```
-> to the agent running on localost?

acidmoose replied:
done

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,190 @@
+import org.secondbase.core.config.SecondBaseModule;
+
+/**
+ * Consul module for SecondBase. Handles registration of services in Consul.
```
Use "consul" instead here. I have a feeling it's mostly used with a lowercase. Super important, i know.

acidmoose replied:
done

neumayer commented on consul/src/test/java/org/secondbase/consul/ConsulModuleIT.java:
```diff
@@ -0,0 +1,115 @@
+import org.junit.Test;
+import org.secondbase.core.SecondBaseException;
+
+public final class ConsulModuleIT {
```
Call ConsulModuleTest?

acidmoose replied:
But it's an integration test which requires a running consul server.

neumayer replied:
Ah, ok.

Commit:
Comment changes and version management for jackson.

- Jackson and Guava needs management because ordering of mvn
dependencies can cause the wrong version to be loaded, and the
wrong version means that (in our case) registering a service
in consul silently fails. Jackson 2.7.4 and Guava 19.0 seem to
work fine.

Commit:
Comment fixes

neumayer commented on example/src/main/java/org/secondbase/example/main/HelloAll.java:
```diff
@@ -0,0 +1,58 @@
+        final ConsulModule consul = new ConsulModule();
+        final SecondBaseModule webconsole = new HttpWebConsole(widgets);
+        final SecondBaseModule jsonLogger = new JsonLoggerModule();
+        final SecondBaseModule[] modules = {consul, webconsole, jsonLogger};
```
The jsonlogger should really be first here since it will define how the consul and webconsole logs look like.

acidmoose replied:
done

neumayer commented on example/src/main/java/org/secondbase/example/main/HelloAll.java:
```diff
@@ -0,0 +1,58 @@
+        // Register custom service in consul
+        // TODO: This should be done automatically by the consul module. Moving common flags to
+        // SecondBase and using those flags from the init methods would allow this.
+        consul.registerServiceInConsul("HelloAll", 8000, "testing", "/healthz", 5L, "testtag");
```
Maybe those should be flags in the consul module and the actual registration should work in the init module there?

Commit:
Make consul module register a service through Flags definitions

Commit:
Logger module first

acidmoose commented on example/src/main/java/org/secondbase/example/main/HelloAll.java:
```diff
@@ -0,0 +1,64 @@
+
+    public static void main(final String[] args) throws SecondBaseException, IOException {
+
+        final String[] realArgs = {
```
As you can see, the arguments passed to the different modules are some of the same: "service" and "environment" at least. We should refactor a bit so that they are set once, in the SecondBase core. Then we can make the init() method of SecondBaseModule pass the SecondBase object, so that modules can optionally use the flags set there.

I'll look into that tomorrow.

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,212 @@
+    /**
+     * Register the service endpoint in consul.
+     *
+     * @param serviceName            The name of the service.
```
No dots here for the params and start with a lowercase letter.

neumayer replied:
And the @throws.

acidmoose replied:
done

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,212 @@
+     * Register the service endpoint with a specific service ID in consul.
+     *
+     * @param serviceName The name of the service.
+     * @param servicePort The port of the service.
```
Also here.

acidmoose replied:
done

neumayer commented on consul/src/main/resources/META-INF/services/org.secondbase.core.config.SecondBaseModule:
```diff
@@ -0,0 +1 @@
+org.secondbase.consul.ConsulModule
```
We don't need that one any more, do we?

acidmoose replied:
done

neumayer commented on consul/src/test/java/org/secondbase/consul/ConsulModuleIT.java:
```diff
@@ -0,0 +1,115 @@
+            assertNotEquals(serviceName, serviceEntry.getService());
+        }
+    }
+}
```
Newline missing.

acidmoose replied:
done

neumayer commented on example/src/main/java/org/secondbase/example/main/HelloConsul.java:
```diff
@@ -0,0 +1,32 @@
+        ConsulModuleConfiguration.host = "localhost:8500";
+
+        // Manually register a service in consul
+        new ConsulModule(ConsulModule.createLocalhostConsulClient()).registerServiceInConsul(
```
This line could just be in the init method, no? In the module. Wouldnt' that just be nice?

acidmoose replied:
It's actually done in the constructor, so I removed this.

Commit:
Remove service description, comment fixing, remove unnecessary stuff from example

Commit:
Checkstyle fixes

neumayer commented on example/src/main/java/org/secondbase/example/main/HelloAll.java:
```diff
@@ -0,0 +1,66 @@
+        final Widget prometheusWidget = new PrometheusWebConsole();
+        final Widget[] widgets = {prometheusWidget};
+
+        final ConsulModule consul = new ConsulModule();
```
This can be of the interface type as well.

acidmoose replied:
done

neumayer commented on example/src/main/java/org/secondbase/example/main/HelloAll.java:
```diff
@@ -0,0 +1,66 @@
+
+        final ConsulModule consul = new ConsulModule();
+        final SecondBaseModule webconsole = new HttpWebConsole(widgets);
+        final SecondBaseModule[] modules = {consul, webconsole, jsonLogger};
```
The jsonlogger should be first so the logging from the webconsole is done in json too.

acidmoose replied:
done

neumayer commented on example/src/main/java/org/secondbase/example/main/HelloAll.java:
```diff
@@ -0,0 +1,66 @@
+
+        mycounter.inc(counter);
+
+        LOG.info(var);
```
I'm not sure about the logic here when it comes to enabling consul. I.e. how to use variables properly. If no service name is specified nothing is happening but if no consul is available we get an error.

acidmoose replied:
You can start a ConsulModule without registering a service through Flags.

ConsulModules c = new ConsulModule();
consulModule.registerService(...);

Commit:
More fixes

neumayer commented on consul/src/main/java/org/secondbase/consul/ConsulModule.java:
```diff
@@ -0,0 +1,208 @@
+
+    @Override
+    public void init() throws SecondBaseException {
+        if (!ConsulModuleConfiguration.serviceName.isEmpty()
```
This happens too quietly. Whether or not the service is registered one can only see in the consul ui. We need some logging for some or all of the cases. "Registered service x at consul y at port a and health path b" or something like that.

acidmoose replied:
Added more logging now.

Commit:
Add logging to consul registration process

Commit:
Checkstyle fix
